src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs

```diff
@@ -159,42 +176,58 @@ private void OnProcessExited(object sender, EventArgs e)
 
             if (Disposing)
             {
-                // No action needed
                 return;
             }
 
             try
             {
-                if (Process.ExitCode == WorkerConstants.SuccessExitCode)
-                {
-                    Process.WaitForExit();
-                    Process.Close();
-                }
-                else if (Process.ExitCode == WorkerConstants.IntentionalRestartExitCode)
+                ThrowIfExitError();
+
+                Process.WaitForExit();
+                if (Process.ExitCode == WorkerConstants.IntentionalRestartExitCode)
                 {
                     HandleWorkerProcessRestart();
                 }
-                else
-                {
-                    string exceptionMessage = string.Join(",", _processStdErrDataQueue.Where(s => !string.IsNullOrEmpty(s)));
-                    string sanitizedExceptionMessage = Sanitizer.Sanitize(exceptionMessage);
-                    var processExitEx = new WorkerProcessExitException($"{Process.StartInfo.FileName} exited with code {Process.ExitCode} (0x{Process.ExitCode.ToString("X")})", new Exception(sanitizedExceptionMessage));
-                    processExitEx.ExitCode = Process.ExitCode;
-                    processExitEx.Pid = Process.Id;
-                    HandleWorkerProcessExitError(processExitEx);
-                }
+            }
+            catch (WorkerProcessExitException processExitEx)
+            {
+                _processExit.TrySetException(processExitEx);
+                HandleWorkerProcessExitError(processExitEx);
             }
             catch (Exception exc)
             {
-                _workerProcessLogger?.LogDebug(exc, "Exception on worker process exit. Process id: {processId}", Process?.Id);
                 // ignore process is already disposed
+                _processExit.TrySetException(exc);
+                _workerProcessLogger?.LogDebug(exc, "Exception on worker process exit. Process id: {processId}", Process?.Id);
             }
             finally
             {
+                _processExit.TrySetResult(Process.ExitCode);
                 UnregisterFromProcessMonitor();
+                Process.Close();
```

Comment:
Non-E2E test step in CI is hanging because some child process is inheriting STDIO and leaving it open. Looking at my PR, the changes to process management seem like the most likely culprit. But I don't understand why -- we were never disposing/closing the process in all cases before. Wouldn't this be a leak in the process restart scenario?